### PR TITLE
Added support for basic Hierarchical Deterministic Key Derivation

### PR DESCRIPTION
**Pull request by erasmospunk**
Added support for hard derivations using the schnorrkel's algorithm.

### CODE REVIEW

File: crypto/src/key/hdkd/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/mod.rs:151]: please use try_fold

File: crypto/src/key/hdkd/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/mod.rs:54]: Like discussion in the call, index can mean index and can mean index | HARDENED_BIT, please disambiguate them.

File: crypto/src/key/hdkd/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/mod.rs:47]: Please remove Hash. It's not needed.

File: crypto/src/key/hdkd/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/mod.rs:30]: Please avoiding such directives

File: crypto/src/key/hdkd/mod.rs
Comment (erasmospunk) [reply crypto/src/key/hdkd/mod.rs]: Done

File: crypto/src/key/hdkd/derivation_path.rs
Comment (iljakuklic) [inline crypto/src/key/hdkd/derivation_path.rs:246]: Maybe tests for `"m/2147483647'"` and `m/2147483648'` should be added

**Review by iljakuklic — COMMENTED** [review]
Adding some comments. Can't say much about security implications.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/derivation_path.rs:72]: Please treat these strictly as ascii bytes. Like I mentioned before, we should not support any unicode or other weird encoding string that works just by coincidence. There's no reason to support it. This is a recipe for problems. This whole wallet system is based on ascii. Let's try to follow that for simplicity.

The correct way to do this is to start by converting this to bytes, check that all the bytes are ascii, and only then start processing.

I do recognize that you added tests for specific cases, but I don't think we can test all possibilities because encoding is a mess in general.

I don't know if I'm missing something here that makes all this OK. But I have to be cautious.

File: crypto/src/key/rschnorr/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/rschnorr/mod.rs:137]: This is not really idiomatic rust. We shouldn't have struct impl blocks outside their modules.

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:107]: The problem with this assumption is that we're saying that *every* kind of PrivateKey now, and in the future, must support being derivable. Not sure this is a good assumption, or even a tenable one.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (iljakuklic) [reply crypto/src/key/hdkd/derivation_path.rs:72]: As far as I can tell, the parsing algorithm already only accepts ASCII strings. The trick is to specify what is accepted, rather than what is rejected, which already happens. The only possible exception might be third-party code like `u32::parse` which is used here. I expect it only accepts ASCII too but may be worth double checking.

Comment (iljakuklic) [reply crypto/src/key/hdkd/derivation_path.rs:72]: I just checked with the stdlib implementation of `u32::parse` and it only accepts ASCII chars. However, it seems it still allows the plus sign even in front of an unsigned number. That means we would accept strings like `"m/+42'"` (please check). Do we want this?

Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/derivation_path.rs:72]: Nice catch! Nope, + isn't allowed.

Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/derivation_path.rs:72]: One thing that also occurred to me... what about exponents? would something like 1e6 parse and return 1000000? That would be very ridiculous! 😄

Quite honestly, this is why I'm overly cautious with parsing strings that are meant to be read as bytes... there's always that corner case that you didn't think of... always some exotic case that'll cause trouble.

In C++, I would use `isdigit()` to check every character after the suffix as bytes. I'm sure there's an equivalent thing in Rust.

Comment (iljakuklic) [reply crypto/src/key/hdkd/derivation_path.rs:72]: > One thing that also occurred to me... what about exponents? would something like 1e6 parse and return 1000000? That would be very ridiculous! smile

Heh. No, from what I've seen, only `+` was there. Also, it may allow leading zeroes.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (erasmospunk) [reply crypto/src/key/hdkd/derivation_path.rs:73]: Not against that

File: crypto/src/key/hdkd/derivation_path.rs
Comment (erasmospunk) [reply crypto/src/key/hdkd/derivation_path.rs:40]: Did you mean checking these errors at run time?

Using a `u31` type will push the checking the result in to that type. I took the approach that @TheQuantumPhysicist suggested, wrapping a private enum in a struct:

```rust
#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
pub struct ChildNumber(DerivationType);

#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
enum DerivationType {
    Hardened(u32),
}
```

So this was it would be impossible to create a wrong kind of ChildNumber i.e. with the internal `u32` having the MSB set to 1. It is not exactly what you were referring to but let me know if this approach is better.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (erasmospunk) [reply crypto/src/key/hdkd/derivation_path.rs:81]: Thank you for this. Originally I wanted to write the code like that but I didn't know that it was possible to pass an array to `strip_suffix`.

File: crypto/src/key/hdkd/mod.rs
Comment (erasmospunk) [reply crypto/src/key/hdkd/mod.rs:36]: Thanks for that, maybe it was a remnant from a previous version.

Regarding `self` vs `&self`, what are the pros and cons of each approach?

File: crypto/src/key/rschnorr/mod.rs
Comment (erasmospunk) [reply crypto/src/key/rschnorr/mod.rs:137]: Would just having a simple function `child_num_to_schnorrkel_chaincode` outside of an `impl` enough?

File: crypto/src/key/mod.rs
Comment (erasmospunk) [reply crypto/src/key/mod.rs:107]: That is correct. The assumption is that because `derive_child` is fallible it would not create any issues.

Another approach would be to have a new private key type `ExtendedPrivateKey` that behaves like a `PrivateKey`.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (erasmospunk) [reply crypto/src/key/hdkd/derivation_path.rs:72]: The idea was that the `parse` to `u32` would fail on non ascii characters. Having said that, I don't have any issues with checking if the `str` is ascii. Better safe than sorry.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (iljakuklic) [inline crypto/src/key/hdkd/derivation_path.rs:113]: Minor: I'd prefer this to be a fully qualified name `std::slice::Iter` rather than importing `Iter` into the module namespace. Just to be more explicit, `Iter` is quite an ambiguous name.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (iljakuklic) [inline crypto/src/key/hdkd/derivation_path.rs:79]: FYI: We have a macro `utils::ensure!` for this pattern.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (iljakuklic) [inline crypto/src/key/hdkd/derivation_path.rs:95]: `HARDENED_H` and `HARDENED_APOS` could be used here

File: crypto/src/key/hdkd/mod.rs
Comment (iljakuklic) [reply crypto/src/key/hdkd/mod.rs]: > Regarding `self` vs `&self`, what are the pros and cons of each approach?

The tradeoffs here may be quite subtle. With `&self` you avoid some cloning/copying but you `.clone()` explicitly here so this advantage is not as pronounced in this case. My motivation for using `self` here is to avoid the initial `.clone()` and also that the previous iteration produces `self` which can be fed into the next iteration so you already have a value there. Also it makes your trait more general. Since you don't require `T: Clone`, it is in principle implementable for more types. I am not sure if that is useful since keys are just byte arrays that tend to be trivially clonable. However, I can imagine a hypothetical scenario where we deliberately avoid implementing `Clone` on some key-like type to let Rust help us to avoid a key reuse when it is undesirable.

Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/mod.rs]: Usually in rust, you're explicitly cloning, in the sense that you force the user of the library to clone in the implementation, then that option can be given to the user. Using self here is a good example of how that works. If the user wants to clone, they can clone the outer object instead of consuming this one.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/derivation_path.rs:89]: Do we want to instead implement ToString, and make this just call that?

File: crypto/src/key/hdkd/derivation_path.rs
Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/derivation_path.rs:45]: Just to clarify, there's also a NonZeroU32, NonZeroU64, etc in rust that solve a similar problem. The benefit here is to ensure that invariants are held across many functions and environments. For example, if function A calls B, and B calls C, ideally you'd have to check the u32 parameter in every call, which is cumbersome. But then if you enforce using U31, then you're guaranteed to do this only once by extracting the logic outside of these functions.

File: crypto/src/key/rschnorr/mod.rs
Comment (TheQuantumPhysicist) [reply crypto/src/key/rschnorr/mod.rs]: There probably should be a trait for this, which you implement for the schnorrkel key.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (iljakuklic) [reply crypto/src/key/hdkd/derivation_path.rs:89]: The way it's laid out in the standard library, it's the other way around. This is defined:
```rust
impl<T: fmt::Display + ?Sized> ToString for T
```

Meaning if you implement `fmt::Display`, you get `ToString` for free.

Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/derivation_path.rs:89]: Nice! Didn't know that! Cheers!

File: crypto/src/key/hdkd/derivation_path.rs
Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/derivation_path.rs:85]: Let's skip that because I wanna avoid adding utils to crypto.

Comment (iljakuklic) [reply crypto/src/key/hdkd/derivation_path.rs:89]: Ok.

**Review by iljakuklic — APPROVED** [review]
This now looks good to me, in terms of code structure and general design. I am not a cryptography expert by any stretch so please wait for an approval by @TheQuantumPhysicist and/or @muursh.

**Review by TheQuantumPhysicist — APPROVED** [review]

**Pull request merged**
